client/my-sites/controller.js

```diff
@@ -235,44 +246,78 @@ module.exports = {
+					sites.once( 'change', () => sites.fetched ? resolve() : reject() );
+				} else {
+					resolve();
+				}
+			} );
+		};
 
-			// if there was a redirect, we should terminate processing of next routes
-			// and let the redirect proceed
-			if ( ! selectionComplete ) {
-				return;
+		const hardFetchSites = () => new Promise( ( resolve, reject ) => {
+			sites.initialized = false;
+			sites.once( 'change', () => {
+				if ( sites.fetched ) {
+					sites.initialized = true;
+					return resolve();
+				}
+
+				reject();
+			} );
+
+			sites.fetch();
+		} );
+
+		let waitingNotice;
+		const validateAllVisibileSitesAreFetched = () => new Promise( resolve => {
+			if ( currentUser.visible_site_count !== sites.getVisible().length ) {
+				if ( ! waitingNotice ) {
+					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
+				}
+
+				return hardFetchSites().then(
+					validateAllVisibileSitesAreFetched,
+					validateAllVisibileSitesAreFetched
+				);
 			}
-		} else {
-			// if sites has fresh data and siteID is invalid
-			// redirect to allSitesPath
-			if ( sites.fetched || ! sites.fetching ) {
-				return page.redirect( allSitesPath );
+
+			if ( waitingNotice ) {
+				notices.removeNotice( waitingNotice );
 			}
 
-			let waitingNotice;
-			const selectOnSitesChange = () => {
-				// if sites have loaded, but siteID is invalid, redirect to allSitesPath
-				if ( sites.select( siteID ) ) {
-					sites.initialized = true;
-					onSelectedSiteAvailable( context );
-					if ( waitingNotice ) {
-						notices.removeNotice( waitingNotice );
-					}
-				} else if ( ( currentUser.visible_site_count !== sites.getVisible().length ) ) {
-					sites.initialized = false;
-					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
-					sites.once( 'change', selectOnSitesChange );
-					sites.fetch();
-				} else {
-					page.redirect( allSitesPath );
+			resolve();
+		} );
+
+		// ensure we have fetched sites
+		const sitesFetchedPromise = waitForInProgressSitesFetch().then( validateAllVisibileSitesAreFetched );
+
+		// Select a site
+		sitesFetchedPromise
+			.catch( error => {
+				page.redirect( allSitesPath );
+				return Promise.reject( error );
+			} )
+			.then( () => {
+				if ( ! sites.select( siteID ) ) {
+					return page.redirect( allSitesPath );
 				}
-			};
-			// Otherwise, check when sites has loaded
-			sites.once( 'change', selectOnSitesChange );
-		}
+
+				const selectionContext = {
+					reactContext: context,
+					selectedSite: sites.getSelectedSite()
+				};
+
+				feedReduxStoreWithSelectedSite( selectionContext );
+
+				if ( isSelectedSiteDomainOnlySite( context.store.getState() ) && ! isPathAllowedForDomainOnlySite( selectionContext ) ) {
+					return renderSelectedSiteIsDomainOnly( selectionContext )
+				}
+
+				setRecentSitesPreferenceInReduxStore( selectionContext );
+			} );
+
 		next();
```

Comment:
The `next()` here is problematic, but I don't currently see how to solve it.

Consider the example of reaching some page that has a loading state, like stats, or post page,
they display a loading state while loading, the controller that displays that "loading state" is the one
that is being called here via `next()`, the thing is, that in certain cases, like if the site is a domain-only site, we want to display the "feature gate", that will promote the features of the full site, however we only know that after sites is loaded, so we get the loading screen and then the feature gate 😞 


posts example:
![posts-loading](https://cloud.githubusercontent.com/assets/326402/22458272/626ff9e2-e7a4-11e6-85b2-18b28bc92e77.gif)


stats example:
![stats-loading](https://cloud.githubusercontent.com/assets/326402/22458298/7de656d0-e7a4-11e6-9281-7f006fdd70b5.gif)